### PR TITLE
Detect and report a FailedUnschedulable VM status

### PR DESCRIPTION
**Pull request by zcahana**
**What this PR does / why we need it**:

This PR extends the VM status field (`.status.printableStatus`) with a new `FailedUnschedulable` status, reported when an error occurs while scheduling the virt-launcher pod for the VM.

This will typically indicate that the VM was configured with unsatisfiable resource requests (e.g., too much memory requested), or unsatisfiable scheduling constraints (e.g., invalid node selector, missing tolerations, ...).

**Special notes for your reviewer**:

Related PR inflight: #6078

**Release note**:
```release-note
Report FailedUnschedulable VM status when scheduling errors occur
```

### CODE REVIEW

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 1]
```diff
@@ -1215,6 +1215,9 @@ const (
 	// VirtualMachineStatusUnknown indicates that the state of the virtual machine could not be obtained,
 	// typically due to an error in communicating with the host on which it's running.
 	VirtualMachineStatusUnknown VirtualMachinePrintableStatus = "Unknown"
+	// VirtualMachineStatusUnschedulable indicates that an error has occurred while scheduling the virtual machime,
+	// e.g. due to unsatisfiable resource requests or unsatisfiable scheduling constraints.
+	VirtualMachineStatusUnschedulable VirtualMachinePrintableStatus = "FailedUnschedulable"
```
Comment (iholder-redhat) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:1220]: Maybe simple `Unschedulable`? or `FailedScheduling`?

File: tests/vm_test.go [hunk 2]
```diff
@@ -693,6 +693,36 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
+				return updatedVm.Status.PrintableStatus
+			}
+
+			By("Verifying that the VM status eventually gets set to FailedUnschedulable")
+			Eventually(vmPrintableStatus, 300*time.Second, 1*time.Second).
+				Should(Equal(v1.VirtualMachineStatusUnschedulable))
+		},
+			table.Entry("unsatisfiable resource requirements", func(vmi *v1.VirtualMachineInstance) {
+				vmi.Spec.Domain.Resources.Requests = corev1.ResourceList{
+					// This may stop working sometime around 2040
```
Comment (iholder-redhat) [inline tests/vm_test.go:714]: 😂

**Review by iholder-redhat — COMMENTED** [review]
Thanks @zcahana! Looks great!

File: tests/vm_test.go [hunk 3]
```diff
@@ -693,6 +693,36 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 				Should(Equal(k8sv1.ConditionFalse))
 		})
 
+		table.DescribeTable("[test_id:todo]should report an error status when VM scheduling error occurs", func(unschedulableFunc func(vmi *v1.VirtualMachineInstance)) {
```
Comment (iholder-redhat) [inline tests/vm_test.go:696]: Don't forget `[test_id:todo]`

Comment (zcahana) [reply tests/vm_test.go:696]: I've requested a test_id to be allocated, waiting for it to arrive.
/hold

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 1 above)
Comment (zcahana) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1220]: I think that all of these options are viable.

- The `Failed` prefix somewhat implies that this failure has occurred while attempting to do something (starting the VM). I think that `Unschedulable` somewhat implies that this is more of a universal state that can always be discovered, whereas in practice we can only discover it while the virt-launcher pod is created and fails to be scheduled.
- The virt-launcher pod fires an event with reason `FailedScheduling`. The VMI sets the `PodScheduled=False` condition with reason `Unschedulable`. I kinda mixed these two to form the `FailedUnschedulable`.

In summary I'm open to all of these, don't have a strong opinion towards any of them. If someone else does, I'm good with changing accordingly :-)

File: tests/vm_test.go (hunk 3 above)
Comment (zcahana) [reply tests/vm_test.go:696]: @ILpinto can you assist in getting a test_id for this?

**Review by stu-gott — COMMENTED** [review]
/approve

**Comment by stu-gott** [comment]
@zcahana just a reminder you put a hold on this PR.

**Comment by zcahana** [comment]
> @zcahana just a reminder you put a hold on this PR.

Right, that was for adding a test_id. I'll just follow up with another PR to complete the missing test_id's once they're ready.

/hold cancel

File: tests/vm_test.go (hunk 3 above)
Comment (kbidarkar) [reply tests/vm_test.go:696]: I will be raising a PR with the test_id for this.

Comment (zcahana) [reply tests/vm_test.go:696]: @kbidarkar that would be great, thanks! :-)
